### PR TITLE
Exclude zero-variance groups from GRPO loss denominator and increase your accuracy

### PR DESCRIPTION
**Pull request by SushantDaga**
Adds a new opt-in GRPOConfig.loss_type value, `dapo_zv`, identical to `dapo` except that sequences belonging to a zero-variance group (every completion in the group received the identical reward, so the group's advantage is exactly 0 for everyone) are excluded from the loss denominator. The numerator is unchanged: those rows already contribute exactly 0 to it. Default ("dapo" and every other existing loss_type) is unchanged.

Implements the loss-mask idea from AGPO (arXiv:2503.15952), applied here only as the zero-advantage-exclusion piece (not AGPO's broader adaptive-loss / token-efficiency proposal).

# What does this PR do?
The denominator overcounts in many RLVR settings where there is a possibility of dead group. This PR implements a solution for it. My literature survey discovered that the same solution has been part of AGPO paper too. In my own experiments I saw better gradient dynamics and increase in accuracy when compared fairly against control and over multiple seeds.

A quick look at some other approved PRs, it seems that adding a separate loss type was a better choice. If you prefer a config change of adding a flag that affects existing losses, do let me know. I am willing to give it some time.

Fixes # (issue)

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Medium Risk**
> Touches core GRPO loss normalization and advantage batching; mistakes could bias gradients, though default paths are untouched and behavior is heavily tested.
> 
> **Overview**
> Adds an opt-in **`GRPOConfig.loss_type="dapo_zv"`** that matches **`dapo`** policy loss and clipping, but normalizes by **`num_items_in_batch_zv`**—active completion tokens from groups where rewards are not all tied—so “dead” zero-advantage groups no longer shrink the effective learning rate. Zero-variance rows are detected from per-group reward min/max after advantages are computed; **`use_liger_kernel=True`** raises **`NotImplementedError`** at trainer init.
> 
> Existing loss types and default **`dapo`** behavior are unchanged. Docs add an AGPO paper index entry; tests cover parity with **`dapo`** when no dead groups, exact denominator rescaling, training smoke, and the Liger guard.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c17988dead62e0feab6f6191d19b58099b691ae3. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -2581,6 +2593,22 @@ def _generate_and_score_completions(
...
+        # when `scale_rewards` is `"group"` or `"none"`; under `scale_rewards="batch"` it instead answers "is the
+        # WHOLE BATCH zero-variance," which would silently degrade `dapo_zv` to plain `dapo` under that mode. This
+        # check is independent of `scale_rewards` and of `multi_objective_aggregation`, since `rewards` (the
+        # already-combined, pre-slice, per-row reward) has the same shape and grouping in both aggregation branches
+        # above. A group containing an unscorable (NaN-reward) row is conservatively treated as NOT zero-variance
+        # (NaN comparisons are always False), which matches plain "dapo" behavior for that row.
+        row_is_zero_variance = (
+            (rewards.view(-1, num_generations).amax(dim=1) - rewards.view(-1, num_generations).amin(dim=1)).abs()
+            <= 1e-6
+        ).repeat_interleave(num_generations)
```
[the remark this answers is not in the document]
Comment (SushantDaga) [reply trl/trainer/grpo_trainer.py:2610]: This seems like a false positive. Each reward is normalized within its own group. So, zero variance groups becomes exactly zero. But also, since all the groups in batch went through same normalization, the batch mean is always zero. So the group's advantage is 0 there too. Look from line L:2579
